### PR TITLE
HTBHF-437 Added address to the claimant object.

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/AddressDTO.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+
+import static uk.gov.dhsc.htbhf.claimant.regex.PostcodeRegex.UK_POST_CODE_REGEX;
+
+@Data
+@Builder
+@AllArgsConstructor(onConstructor_ = {@JsonCreator})
+@ApiModel(description = "An address for help to buy healthy foods.")
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/AddressDTO.java]: Don't think this is a great description, should say what it is used for.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/AddressDTO.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    @NotNull
+    @Size(min = 1, max = 500)
+    @JsonProperty("addressLine1")
+    @ApiModelProperty(notes = "First line of the card delivery address", example = "Flat B")
+    private String addressLine1;
+
+    @Size(max = 500)
+    @JsonProperty("addressLine2")
+    @ApiModelProperty(notes = "Second line of the card delivery address", example = "123 Fake Street")
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/model/AddressDTO.java]: Minor - Probably shouldn't use the work `Fake`.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/converter/AddressDTOToAddressConverter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+import uk.gov.dhsc.htbhf.claimant.entity.Address;
+import uk.gov.dhsc.htbhf.claimant.model.AddressDTO;
+
+/**
+ * Converts a {@link AddressDTO} into a {@link Address}.
+ */
+@Component
+public class AddressDTOToAddressConverter {
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/converter/AddressDTOToAddressConverter.java]: You need a unit test for this class.

File: api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy [hunk 4]
```diff
@@ -140,6 +145,13 @@ class NewClaimSpec extends Specification {
         assertValidationResponse(response, "claimant.dateOfBirth", "must not be null")
     }
 
+    private assertAddress(Address actual, AddressDTO expected) {
```
Comment (mikeSmith08) [inline api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy]: Not a great name for the method, doesn't say what it's asserting, should at least say `assertAddressValid` or similar.

File: api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy [hunk 5]
```diff
@@ -140,6 +145,13 @@ class NewClaimSpec extends Specification {
...
+        assertThat(actual.addressLine1).isEqualTo(expected.addressLine1)
```
Comment (mikeSmith08) [inline api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy]: I know it will fail if this is the case, but I always like to have a specific assertion to make sure the object is not null first so we can control the error message better.

File: api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy [hunk 6]
```diff
@@ -103,6 +88,26 @@ class NewClaimSpec extends Specification {
...
+        fieldName      | value       | expectedErrorMessage             | expectedField
+        "addressLine1" | null        | "must not be null"               | "claimant.cardDeliveryAddress.addressLine1"
+        "addressLine1" | LONG_STRING | "size must be between 1 and 500" | "claimant.cardDeliveryAddress.addressLine1"
+        "addressLine2" | LONG_STRING | "size must be between 0 and 500" | "claimant.cardDeliveryAddress.addressLine2"
+        "townOrCity"   | null        | "must not be null"               | "claimant.cardDeliveryAddress.townOrCity"
+        "townOrCity"   | LONG_STRING | "size must be between 1 and 500" | "claimant.cardDeliveryAddress.townOrCity"
+        "postcode"     | "AA1122BB"  | "invalid postcode format"        | "claimant.cardDeliveryAddress.postcode"
+        "postcode"     | "A"         | "invalid postcode format"        | "claimant.cardDeliveryAddress.postcode"
+        "postcode"     | "11AA21"    | "invalid postcode format"        | "claimant.cardDeliveryAddress.postcode"
+        "postcode"     | ""          | "invalid postcode format"        | "claimant.cardDeliveryAddress.postcode"
```
Comment (mikeSmith08) [inline api/src/test/groovy/uk/gov/dhsc/htbhf/claimant/controller/NewClaimSpec.groovy]: Can we have a test (probably needs to be separate from this test) for a null postcode please?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/ClaimDTOTestDataFactory.java [hunk 7]
```diff
@@ -37,7 +42,16 @@ public static ClaimDTO aClaimDTOWithDateOfBirth(LocalDate dateOfBirth) {
                 .firstName(VALID_FIRST_NAME)
                 .lastName(VALID_LAST_NAME)
                 .nino(VALID_NINO)
-                .dateOfBirth(VALID_DOB);
+                .dateOfBirth(VALID_DOB)
+                .cardDeliveryAddress(aValidAddressBuilder().build());
+    }
+
+    private static AddressDTO.AddressDTOBuilder aValidAddressBuilder() {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/testsupport/ClaimDTOTestDataFactory.java]: I think this would be better in its own TestDataFactory.

File: swagger.yml [hunk 8]
```diff
@@ -48,6 +48,35 @@ paths:
...
+        description: "Second line of the card delivery address"
+        minLength: 0
+        maxLength: 500
+      postcode:
+        type: "string"
+        example: "AA1 1AA"
+        description: "The postcode to send the address to."
+        pattern: "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\\
+          s?[0-9][A-Za-z]{2})"
+      townOrCity:
```
Comment (mikeSmith08) [inline swagger.yml]: Minor - Do we need to reorder this so that postcode is last?
